compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -17,66 +19,122 @@ module Development.IDE.Types.Diagnostics (
   errorDiag,
   ideTryIOException,
   prettyFileDiagnostics,
-  prettyDiagnostic
+  prettyDiagnostic,
+  defDiagnostic,
+  addDiagnostics,
+  addLocation,
+  addFilePath
   ) where
 
-import Control.DeepSeq
 import Control.Exception
+import qualified Control.Lens as L
 import Data.Aeson (FromJSON, ToJSON)
 import Data.Either.Combinators
 import Data.List.Extra
+import Data.Maybe as Maybe
 import qualified Data.Text as T
 import Data.Text.Prettyprint.Doc.Syntax
 import GHC.Generics
-import qualified Network.URI.Encode
 import qualified Text.PrettyPrint.Annotated.HughesPJClass as Pretty
+import           Language.Haskell.LSP.Types as LSP (
+    DiagnosticSeverity(..)
+  , Diagnostic(..)
+  , filePathToUri
+  , List(..)
+  , DiagnosticRelatedInformation(..)
+  )
+import Language.Haskell.LSP.Diagnostics
 
 import Development.IDE.Types.Location
 
-ideErrorText :: FilePath -> T.Text -> Diagnostic
-ideErrorText absFile = errorDiag absFile "Ide Error"
-
-ideErrorPretty :: Pretty.Pretty e => FilePath -> e -> Diagnostic
-ideErrorPretty absFile = ideErrorText absFile . T.pack . Pretty.prettyShow
-
-errorDiag :: FilePath -> T.Text -> T.Text -> Diagnostic
-errorDiag fp src msg =
-  Diagnostic
-  { dFilePath = fp
-  , dRange    = noRange
-  , dSeverity = Error
-  , dSource   = src
-  , dMessage  = msg
+ideErrorText :: FilePath -> T.Text -> LSP.Diagnostic
+ideErrorText fp = errorDiag fp "Ide Error"
+
+ideErrorPretty :: Pretty.Pretty e => FilePath -> e -> LSP.Diagnostic
+ideErrorPretty fp = ideErrorText fp . T.pack . Pretty.prettyShow
+
+errorDiag :: FilePath -> T.Text -> T.Text -> LSP.Diagnostic
+errorDiag fp src =
+  addFilePath fp . diagnostic noRange LSP.DsError src
+
+-- | This is for compatibility with our old diagnostic type
+diagnostic :: Range
+           -> LSP.DiagnosticSeverity
+           -> T.Text -- ^ source
+           -> T.Text -- ^ message
+           -> LSP.Diagnostic
+diagnostic rng sev src msg
+    = LSP.Diagnostic {
+          _range = rng,
+          _severity = Just sev,
+          _code = Nothing,
+          _source = Just src,
+          _message = msg,
+          _relatedInformation = Nothing
+          }
+
+-- | Any optional field is instantiated to Nothing
+defDiagnostic ::
+  Range ->
+  T.Text -> -- ^ error message
+  LSP.Diagnostic
+defDiagnostic _range _message = LSP.Diagnostic {
+    _range
+  , _message
+  , _severity = Nothing
+  , _code = Nothing
+  , _source = Nothing
+  , _relatedInformation = Nothing
   }
 
-ideTryIOException :: FilePath -> IO a -> IO (Either Diagnostic a)
+-- | addLocation but with no range information
+addFilePath ::
+  FilePath ->
+  LSP.Diagnostic ->
+  LSP.Diagnostic
+addFilePath fp =
+  addLocation $ Location (filePathToUri fp) noRange
+
+-- | This adds location information to the diagnostics but this is only used in
+--   the case of serious errors to give some context to what went wrong
+addLocation ::
```

Comment:
Ditto, setLocation.

Reply:
This does accumulate location information (line 110). It does this because we try to detect whether any of our diagnostics originate from outside the current shake target

Reply:
👍